Retrieve the moRef of a datastore

RichardHoch opened the pull request:
### JIRA

* [MTV-1099](https://issues.redhat.com/browse/MTV-1099)

### MTV 2.5.6

Resolves https://issues.redhat.com/browse/MTV-1099 by adding instructions in the CLI migration procedures that explain how a user can retrieve a moRef of a datastore.

Also fixes a few minor formatting issues. 

Previews:
1. New section: https://file.emea.redhat.com/rhoch/mor/html-single/#retrieving-vmware-mor_mtv 
2. Note directing users to new section: https://file.emea.redhat.com/rhoch/mor/html-single/#migrating-virtual-machines-cli_mtv [search for "moRef"]

RichardHoch commented:
@ahadas Please review. This PR is for MTV 2.5.6 _only_.

ahadas commented on documentation/modules/migrating-virtual-machines-cli.adoc:
```diff
@@ -135,8 +135,21 @@ spec:
 ----
 <1> Specify the name of the VMware `Provider` CR.
-<2> Specify the managed object reference (MOR) of the VMware host.
+<2> Specify the managed object reference (MOR) of the VMware host. To retrieve the name of the MOR, see the note that follows.
```
we refer to MOR also in other places, so I think we should have this before

RichardHoch replied:
@ahadas The Host manifest is the first place we refer to MOR, but, of course you're right, we refer to it elsewhere (later).  Do you want me to a) move the note to before the procedure b) include the note for all manifests where MOR  is mentioned c) write "To retrieve the MOR, see..." for the other mentions, without copy-pasting the note.

ahadas replied:
I think that moving it before the procedure makes more sense

RichardHoch replied:
Moved the new section to before the migration procedure.

ahadas commented on documentation/modules/migrating-virtual-machines-cli.adoc:
```diff
@@ -135,8 +135,21 @@ spec:
++
+[NOTE]
+====
+To retrieve the name of the MOR:
```
not sure about "the name of the MOR" since it's not really its name, its the MOR itself
maybe just "To retrieve the MOR of an entity:" ?

RichardHoch replied:
Fixed.

ahadas commented on documentation/modules/migrating-virtual-machines-cli.adoc:
```diff
@@ -135,8 +135,21 @@ spec:
+. Select the `rootFolder`.
+. Select the `childEntity` datacenter object type.
++
+The datastores, including the MOR, are displayed.
```
and then generalize this part as well - we see there the MOR of other entities as well

RichardHoch replied:
@ahadas Do you mean write "The datastores including the MORs, are displayed." ?

ahadas replied:
nope, I meant to replace this with something like "In this screen, the MOR of datastores, networks and other vSphere entities, are displayed.

ahadas replied:
```suggestion
From this screen you can navigate to entities within the datacenter and find their MOR.
```

anarnold97 commented:
Please do not merge.

Highlight at me and i will deal with it. 2.5 in GitHub is not syncing correctly with 2.5 in GitLab

I will have to push this directly to GitLab and not try to sync using the script.  I set up a fork last night so I can get this done quickly enough

RichardHoch commented:
@anarnold97 Don;t worry; i will not merge anything to 2.5 unless you give approval. I'll work on the content on Sunday and we cam talk on Monday about how to best proceed. In version 2.5, where we have only one of set CLI instructions, we'll probably want to have this information in the text before the procedure. In version 2.6, it should go before the VMware (vSphere) set of instructions.

anarnold97 commented:
@RichardHoch - I will be here on Sunday. If you get the 2.5 doc done, just send it through to me and will push it to GitLab

ahadas reviewed (commented):
@RichardHoch while looking at more closely at https://<url_of_vCenter_server_with_VMware_host>/mob I realized that it would be simpler to get this from the inventory service as explain in the [Retrieving the Inventory service JSON](https://access.redhat.com/documentation/en-us/migration_toolkit_for_virtualization/2.6/html-single/installing_and_using_the_migration_toolkit_for_virtualization/index#retrieving-validation-service-json_mtv) section - the id of vSphere entities refers to their MOR in vSphere
for example, when we get the following datastore from the inventory:
```yaml
 {
    "id": "datastore-729",
    "parent": {
      "kind": "Folder",
      "id": "group-s5"
    },
    "path": "/Datacenter/datastore/f01-h27-640-SSD_1",
    "revision": 17,
    "name": "f01-h27-640-SSD_1",
    "selfLink": "providers/vsphere/01278af6-e1e4-4799-b01b-d5ccc8dd0201/datastores/datastore-729"
  }
```
we know that the MOR of datastore "f01-h27-640-SSD_1" is datastore-729

so I think we should recommend this approach and mention the procedure of getting this from vSphere as an alternative

RichardHoch commented:
@ahadas I've taken this PR in a somewhat different direction as we've discussed the last few days. It's ready for you to review again.

RichardHoch commented:
@ahadas I moved the instructions for retrieving the MOR so that they appear before the procedure for migrating via CLI, instead of in the callouts. I added cross-references to the relevant callouts and also fixed the heading level of the Inventory JSON section in the "Troubleshooting."

That should cover everything. Please review the PR again.

ahadas commented on documentation/modules/retrieving-vmware-mor.adoc:
```diff
@@ -0,0 +1,73 @@
++
+[source,terminal]
+----
+$ curl -H "Authorization: Bearer $TOKEN"  https://<inventory_service_route>/providers/<datastore>/<MOR>/vms -k
```
but when we get here, we don't know the MOR yet so we should rather ask the user to list the datastores:
```suggestion
$ curl -H "Authorization: Bearer $TOKEN"  https://<inventory_service_route>/providers/vsphere/<provider id>/datastores/ -k
```
and then find the datastore in the output, for example:
```json
[
  {
    "id": "datastore-11",
    "parent": {
      "kind": "Folder",
      "id": "group-s5"
    },
    "path": "/Datacenter/datastore/v2v_general_porpuse_ISCSI_DC",
    "revision": 46,
    "name": "v2v_general_porpuse_ISCSI_DC",
    "selfLink": "providers/vsphere/01278af6-e1e4-4799-b01b-d5ccc8dd0201/datastores/datastore-11"
  },
  {
    "id": "datastore-730",
    "parent": {
      "kind": "Folder",
      "id": "group-s5"
    },
    "path": "/Datacenter/datastore/f01-h27-640-SSD_2",
    "revision": 46,
    "name": "f01-h27-640-SSD_2",
    "selfLink": "providers/vsphere/01278af6-e1e4-4799-b01b-d5ccc8dd0201/datastores/datastore-730"
  },
 ....
```
and the `id` field would indicate the MOR of the datastore

RichardHoch commented:
@ahadas I made the changes you suggested. Please review this PR again.

anarnold97 commented on documentation/modules/retrieving-vmware-mor.adoc:
```diff
@@ -0,0 +1,83 @@
+[id="retrieving-vmware-mor_{context}"]
+= Retrieving a VMware vSphere MOR
+
+When you migrate VMs with a VMware vSphere source provider using {project-first} from the CLI, you need to know the managed object reference (MOR) of certain entities in vSphere, such as datastores, networks and VMs.
```
Remember to use the Oxford comma in comma separated lists
```suggestion
When you migrate VMs with a VMware vSphere source provider using {project-first} from the CLI, you need to know the managed object reference (MOR) of certain entities in vSphere, such as datastores, networks, and VMs.
```

RichardHoch replied:
Doh! 

Note that Vale didn't catch that... [still on me, though] 
![image](https://github.com/kubev2v/forklift-documentation/assets/60698649/e220ad31-7522-45a4-9a2e-50ee8e14721d)

anarnold97 replied:
NOTE: Vale is a linter for RH style. 
If you need a grammar linter to correct basic grammar errors, I am happy to suggest one to you.

ahadas commented on documentation/modules/retrieving-vmware-mor.adoc:
```diff
@@ -0,0 +1,81 @@
+
+When you migrate VMs with a VMware vSphere source provider using {project-first} from the CLI, you need to know the managed object reference (MOR) of certain entities in vSphere, such as datastores, networks, and VMs.
+
+You can retrieve the MOR of one or more vSphere datastores from the Inventory service. You can then use each MOR as a reference for retrieving the MOR of another entity.
```
I'm not sure about this part - it applies to snapshots since we start from the datastore that the corresponding disks reside on, but if we take virtual machines - they are not placed on a datastore so if we need their MOR, the datastore won't be their entrypoint

I'd suggest to generalize the first sentence to be more general, e.g., `You can retrieve the MOR of one or more vSphere entities from the Inventory service.`

RichardHoch replied:
Fixed.

ahadas reviewed (commented):
one comment inside, besides that lgtm

RichardHoch commented:
@ahadas I made the change you suggested.

ahadas reviewed (approved):
thanks @RichardHoch!

anarnold97 commented on documentation/modules/migrating-virtual-machines-cli.adoc:
```diff
@@ -135,7 +135,7 @@ spec:
 ----
 <1> Specify the name of the VMware `Provider` CR.
-<2> Specify the managed object reference (MOR) of the VMware host.
+<2> Specify the managed object reference (MOR) of the VMware host. To retrieve the MOR, see xref:retrieving-vmware-mor_{context}[Retrieving a VMware vSphere MOR].
```
This xref is wrong. Use the actual context, not the context attribute. It will generate a build error when we move it to Pantheon

anarnold97 replied:
WRONG - apologies

anarnold97 commented on documentation/modules/migrating-virtual-machines-cli.adoc:
```diff
@@ -175,7 +175,7 @@ EOF
 <1> Allowed values are `pod` and `multus`.
 <2> You can use either the `id` _or_ the `name` parameter to specify the source network.
-<3> Specify the VMware network MOR, the {rhv-short} network UUID, or the {osp} network UUID.
+<3> Specify the VMware network MOR, the {rhv-short} network UUID, or the {osp} network UUID. To retrieve the MOR, see xref:retrieving-vmware-mor_{context}[Retrieving a VMware vSphere MOR].
```
And this is wrong as well

anarnold97 commented on documentation/modules/migrating-virtual-machines-cli.adoc:
```diff
@@ -214,7 +214,7 @@ spec:
 ----
 <1> Allowed values are `ReadWriteOnce` and `ReadWriteMany`.
-<2> Specify the VMware data storage MOR, the {rhv-short} storage domain UUID, or the {osp} `volume_type` UUID. For example, `f2737930-b567-451a-9ceb-2887f6207009`.
+<2> Specify the VMware datastore MOR, the {rhv-short} storage domain UUID, or the {osp} `volume_type` UUID. For example, `f2737930-b567-451a-9ceb-2887f6207009`. To retrieve the MOR, see xref:retrieving-vmware-mor_{context}[Retrieving a VMware vSphere MOR].
```
And here

anarnold97 replied:
WRONG - apologies

anarnold97 commented on documentation/modules/migrating-virtual-machines-cli.adoc:
```diff
@@ -291,7 +291,7 @@ EOF
 <8> For all source providers except for {virt}, you can use either the `id` _or_ the `name` parameter to specify the source VMs. +
 {virt} source provider only: You can use only the `name` parameter, not the `id.` parameter to specify the source VMs.
-<9> Specify the VMware VM MOR, {rhv-short} VM UUID or the {osp} VM UUID.
+<9> Specify the VMware VM MOR, {rhv-short} VM UUID or the {osp} VM UUID. To retrieve the MOR, see xref:retrieving-vmware-mor_{context}[Retrieving a VMware vSphere MOR].
```
And here

anarnold97 replied:
WRONG - apologies

anarnold97 commented on documentation/modules/retrieving-vmware-mor.adoc:
```diff
@@ -0,0 +1,81 @@
+$ TOKEN=$(oc whoami -t)
+----
+
+. Retrieve the `MOR` of a VMware vSphere provider:
```
Why the backticks on this MOR but not earlier in the documentation?

Please be consistent with your formatting

anarnold97 commented on documentation/modules/rhv-prerequisites.adoc:
```diff
@@ -11,7 +11,7 @@ The following prerequisites apply to {rhv-full} migrations:
 * You must have the {manager} CA certificate, unless it was replaced by a third-party certificate, in which case, specify the {manager} Apache CA certificate.
 +
-You can obtain the {manager}  CA certificate by navigating to https://<engine_host>/ovirt-engine/services/pki-resource?resource=ca-certificate&format=X509-PEM-CA in a browser.
+You can obtain the {manager} CA certificate by navigating to https://<engine_host>/ovirt-engine/services/pki-resource?resource=ca-certificate&format=X509-PEM-CA in a browser.
```
```suggestion
You can obtain the {manager} CA certificate by navigating to `https://<engine_host>/ovirt-engine/services/pki-resource?resource=ca-certificate&format=X509-PEM-CA` in a browser.
```

RichardHoch replied:
Good catch! Fixed.

RichardHoch commented:
@anarnold97 I made the changes you suggested. Please review again. If good, please merge to the 2.5 branch. I will create a separate PR for the 2.6 branch because the 2.6 documentation has a different structure for the section that describes migration using the CLI.

RichardHoch commented:
@anarnold97  Don't merge this PR. I now see that "MOR" really should be "moRef" -- will fix and notify you.

anarnold97 reviewed (approved):
LGTM - thanks

RichardHoch commented:
@ahadas Please review -- I simply changed "MOR" to "moRef" after you approved the previous version.

RichardHoch commented:
@anarnold97 Please give this a final review -- if good, it's good for inclusion in the main branch. Thanks.

anarnold97 reviewed (approved):
LGTM

anarnold97 commented:
@RichardHoch - when you are happy, please give me a shout. I will handle the merge later today.

RichardHoch commented:
@anarnold97 Please merge and publish to the 2.5 branch of MTV. Thanks.

anarnold97 commented:
@RichardHoch - Pretty sure we have published this to GitLab and published on Monday. Please can you double check, and close if it is?

RichardHoch commented:
@anarnold97 Yes --it's in both the 2.5 and 2.6 documentation set. i will now the Jira with live links. Thanks!

anarnold97 commented:
> @anarnold97 Yes --it's in both the 2.5 and 2.6 documentation set. i will now the Jira with live links. Thanks!

@RichardHoch - please we CLOSE this ticket?
Thanks
